keps/sig-node/4781-kublet-restart-pod-status/README.md

```diff
@@ -0,0 +1,581 @@
+
+## Motivation
+
+Ensuring high availability and minimizing service disruptions are critical considerations for Kubernetes clusters. 
+When the kubelet restarts, it resets the Start and Ready states of all containers to False by default. This means that any successful probe statuses that were previously established are lost upon the restart. As a result, services may be inaccurately flagged as unavailable, despite having been operational prior to the kubelet's restart. This reset can lead to erroneous perceptions of service health and negatively impact the overall performance of the cluster, potentially triggering unnecessary alerts or load balancing changes. 
+It's essential to implement strategies to ensure that the service states accurately reflect their operational status, even during kubelet interruptions.
+
+### Goals
+
+- Ensure consistency in container start and ready states across kubelet restarts.
+- Minimize unnecessary service disruptions caused by temporary ready state changes.
+
+### Non-Goals
+
+- If the kubelet fails to renew its lease beyond the nodeMonitorGracePeriod due to an excessively long restart interval, the Ready status of the containers in the pods on the node will be set to false. In this situation, we should not manually set the Ready status back to true. Instead, it should remain false, waiting for the probe to execute again and restore it.
+- Modify the fundamental logic of how readiness probes work.
+
+## Proposal
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As a user of Kubernetes, I want the container's Ready state to remain consistent across kubelet restarts so that my services do not experience unnecessary downtime.
+However, currently, a kubelet restart causes a brief "Not Ready" storm, where the state of all Pods is set to Not Ready, impacting the availability of my services.
+
+### Risks and Mitigations
+
+##### Inconsistency with other Kubernetes components
+If other parts of Kubernetes (e.g., the API server, controllers) expect certain behavior regarding container readiness states, these changes might cause inconsistencies.
+
+##### Delayed Health Check Updates
+By preserving the old state without immediate health checks, there is a delay in recognizing containers that have become unhealthy during or after kubelet's downtime. Services relying on Pod readiness for service discovery might continue directing traffic to Pods with containers that are no longer healthy but are still reported as Ready.
+We plan to immediately trigger a probe after that to reduce the risk caused by such delays.
```

Comment:
This seems to contradict current approach.
https://github.com/kubernetes/kubernetes/blob/v1.34.0/pkg/kubelet/prober/worker.go#L152-L157

If a Node experiences intermittent issues and the kubelet keeps restarting, having probes run in rapid succession could be problematic. Will this behavior be modified within the KEP (i.e. temporary `deprecated` FeatureGate)?

Reply:
This is a good point.  We should probably do something like:

```
at startup {
    for each running pod
        for each probe
            wait for a random amount of time between 0 and the probe's period
            start periodic probing
        }
    }
}
```

If that is too long we could reduce it to 1/2 the period, or maybe vary that coefficient between "fast" and "slow" probes.